### PR TITLE
DISA K8s STIG rule 242442 does not parse imageID correctly

### LINKED ISSUE #302: DISA K8s STIG rule 242442 does not parse `imageID` correctly
**What happened**:
Currently DISA K8s STIG rule 242442 for all providers is not parsing `imageID` correctly. Example for `gardener` provider:
https://github.com/gardener/diki/blob/00865968061d7c97dbf0968776e20d77045c301d/pkg/provider/gardener/ruleset/disak8sstig/rules/242442.go#L98-L99

As seen the `imageID` is split by `:` and the right half is used for base image. However, it is possible for the base image to contain `:` substring ( example of `imageID`: `foo:1234@sha256:f7ee3d33e8f1795bba370bbbf61e87e97c530f323c9c48aec5a3f1ebcdc41cab`) which would result in a wrongly parsed base image.

**What you expected to happen**:
DISA K8s STIG rule 242442 should correctly parse `imageID`.

### PR DESCRIPTION
**Pull request by georgibaltiev**
**What this PR does / why we need it**:
This PR refactors the parsing of image names in all 3 implementations of DISA rule 242442. The strings split functions are replaced by an regex parsing function imported from the package k8s.io/kubernetes/pkg/util/parsers. An additional test to replicate the scenario that provoked the issue has been added as well.

**Which issue(s) this PR fixes**:
Fixes #302

**Special notes for your reviewer**:

**Release note**:

```bugfix user
A bug causing rule 242442 for all providers to parse image names incorrectly under certain circumstances (for example the image name containing additional semicolons before the tag/digest) was fixed.
```

### CODE REVIEW

File: pkg/provider/gardener/ruleset/disak8sstig/rules/242442.go [hunk 1]
```diff
@@ -96,7 +96,13 @@ func (*Rule242442) checkImages(cluster string, pods []corev1.Pod) []rule.CheckRe
 				}
 
 				imageRef := pod.Status.ContainerStatuses[containerStatusIdx].ImageID
-				imageBase := strings.Split(strings.Split(imageRef, ":")[0], "@")[0]
+				named, err := dockerref.ParseNormalizedNamed(imageRef)
+				if err != nil {
+					checkResults = append(checkResults, rule.ErroredCheckResult(err.Error(), rule.NewTarget("cluster", cluster, "pod", pod.Name, "container", container.Name, "imageRef", imageRef)))
```
Comment (AleksandarSavchev) [inline pkg/provider/gardener/ruleset/disak8sstig/rules/242442.go:101]:
```suggestion
					checkResults = append(checkResults, rule.ErroredCheckResult(err.Error(), rule.NewTarget("imageRef", imageRef)))
```
Nit: No need to be too specific here as we should never enter this `if` clause. If we reach this code it is more likely an error in `ParseNormalizedNamed` and we would only care about the `imageRef` it caused it.

File: pkg/provider/virtualgarden/ruleset/disak8sstig/rules/242442.go [hunk 2]
```diff
@@ -67,7 +67,13 @@ func (*Rule242442) checkImages(pods []corev1.Pod, images map[string]string, repo
 			}
 
 			imageRef := pod.Status.ContainerStatuses[containerStatusIdx].ImageID
-			imageBase := strings.Split(strings.Split(imageRef, ":")[0], "@")[0]
+			named, err := dockerref.ParseNormalizedNamed(imageRef)
+			if err != nil {
+				checkResults = append(checkResults, rule.ErroredCheckResult(err.Error(), rule.NewTarget("name", pod.Name, "container", container.Name, "imageRef", imageRef)))
```
Comment (AleksandarSavchev) [inline pkg/provider/virtualgarden/ruleset/disak8sstig/rules/242442.go:72]:
```suggestion
				checkResults = append(checkResults, rule.ErroredCheckResult(err.Error(), rule.NewTarget("imageRef", imageRef)))
```

File: pkg/provider/gardener/ruleset/disak8sstig/rules/242442_test.go [hunk 3]
```diff
@@ -167,11 +170,37 @@ var _ = Describe("#242442", func() {
 
 		Expect(ruleResult.CheckResults).To(Equal(expectedCheckResults))
 	})
+	It("should return correct results when an image uses more than 1 version #2", func() {
```
Comment (AleksandarSavchev) [inline pkg/provider/gardener/ruleset/disak8sstig/rules/242442_test.go:173]:
```suggestion
	It("should return correct results when images are local", func() {
```
Let's choose a more specific

File: pkg/provider/managedk8s/ruleset/disak8sstig/rules/242442_test.go [hunk 4]
```diff
@@ -88,27 +91,53 @@ var _ = Describe("#242442", func() {
...
 		expectedCheckResults := []rule.CheckResult{
-			rule.FailedCheckResult("Image is used with more than one versions.", rule.NewTarget("kind", "node", "name", "foo", "image", "eu.gcr.io/image2")),
-			rule.FailedCheckResult("Image is used with more than one versions.", rule.NewTarget("kind", "node", "name", "foo", "image", "eu.gcr.io/image3")),
+			rule.FailedCheckResult("Image is used with more than one versions.", rule.NewTarget("kind", "node", "name", "foo", "image", "eu.gcr.io/image-two")),
+			rule.FailedCheckResult("Image is used with more than one versions.", rule.NewTarget("kind", "node", "name", "foo", "image", "eu.gcr.io/image-three")),
+		}
+
+		Expect(ruleResult.CheckResults).To(Equal(expectedCheckResults))
+	})
+	It("should return correct results when an image uses more than 1 version #2", func() {
```
Comment (AleksandarSavchev) [inline pkg/provider/managedk8s/ruleset/disak8sstig/rules/242442_test.go:119]:
```suggestion
	It("should return correct results when images are local", func() {
```

File: pkg/provider/virtualgarden/ruleset/disak8sstig/rules/242442_test.go [hunk 5]
```diff
@@ -78,27 +81,53 @@ var _ = Describe("#242442", func() {
...
 		Expect(err).ToNot(HaveOccurred())
 
 		expectedCheckResults := []rule.CheckResult{
-			rule.FailedCheckResult("Image is used with more than one versions.", rule.NewTarget("image", "eu.gcr.io/image2")),
+			rule.FailedCheckResult("Image is used with more than one versions.", rule.NewTarget("image", "eu.gcr.io/image-two")),
+		}
+
+		Expect(ruleResult.CheckResults).To(Equal(expectedCheckResults))
+	})
+	It("should return correct results when an image uses more than 1 version #2", func() {
```
Comment (AleksandarSavchev) [inline pkg/provider/virtualgarden/ruleset/disak8sstig/rules/242442_test.go:100]:
```suggestion
	It("should return correct results when images are local", func() {
```

File: pkg/provider/gardener/ruleset/disak8sstig/rules/242442_test.go [hunk 6]
```diff
@@ -167,11 +170,37 @@ var _ = Describe("#242442", func() {
...
+		r := &rules.Rule242442{ClusterClient: fakeShootClient, ControlPlaneClient: fakeSeedClient, ControlPlaneNamespace: namespace}
+		pod1 := shootPod.DeepCopy()
+		pod1.Name = "pod1"
+		pod1.Status.ContainerStatuses[0].ImageID = "localhost:7777/image-one@sha256:" + digest1
+		pod1.Status.ContainerStatuses[1].ImageID = "localhost:7777/image-one@sha256:" + digest2
+		pod1.Status.ContainerStatuses[2].ImageID = "localhost:7777/image-two@sha256:" + digest3
```
Comment (AleksandarSavchev) [inline pkg/provider/gardener/ruleset/disak8sstig/rules/242442_test.go:179]:
```suggestion
		pod1.Status.ContainerStatuses[0].ImageID = "localhost:7777/image1@sha256:" + digest1
		pod1.Status.ContainerStatuses[1].ImageID = "localhost:7777/image1@sha256:" + digest2
		pod1.Status.ContainerStatuses[2].ImageID = "localhost:7777/image2@sha256:" + digest3
```
Nit: We should use either `image[1-9]{1}` or `image-[a-z]+` approach for all cases in out unit tests. I prefer `image[1-9]{1}` but I would be ok if you change them all to `image-[a-z]+` as well.

**Review by AleksandarSavchev — CHANGES_REQUESTED** [review]
Looks good overall, just a few nits.

File: pkg/provider/gardener/ruleset/disak8sstig/rules/242442_test.go (hunk 6 above)
Comment (georgibaltiev) [reply pkg/provider/gardener/ruleset/disak8sstig/rules/242442_test.go:179]: The regex will not accept any numbers in the repo/name part of the ImageID. I will use dashes and letters instead

**Review by AleksandarSavchev — COMMENTED** [review]
Thanks!
/lgtm

File: pkg/provider/managedk8s/ruleset/disak8sstig/rules/242442.go [hunk 7]
```diff
@@ -7,8 +7,8 @@ package rules
 import (
 	"context"
 	"slices"
-	"strings"
 
+	dockerref "github.com/distribution/reference"
```
Comment (dimityrmirchev) [inline pkg/provider/managedk8s/ruleset/disak8sstig/rules/242442.go:11]:
```suggestion
	imageref "github.com/distribution/reference"
```

File: pkg/provider/managedk8s/ruleset/disak8sstig/rules/242442_test.go [hunk 8]
```diff
@@ -113,19 +116,45 @@ var _ = Describe("#242442", func() {
 
 		Expect(ruleResult.CheckResults).To(Equal(expectedCheckResults))
 	})
+	It("should return correct results when images are local", func() {
```
Comment (dimityrmirchev) [inline pkg/provider/managedk8s/ruleset/disak8sstig/rules/242442_test.go:119]:
```suggestion
	It("should return correct results when image repository includes port number", func() {
```

**Review by dimityrmirchev — CHANGES_REQUESTED** [review]
Thanks! Two requests that need to be applied in several places. The rest looks good.

**Review by AleksandarSavchev — APPROVED** [review]

**Review by dimityrmirchev — APPROVED** [review]
/lgtm

**Pull request merged**